sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,333 @@
+        final long inactiveTimeLimitInSeconds, final PartitionPumpManager partitionPumpManager) {
+        this.partitionManager = partitionManager;
+        this.eventHubAsyncClient = eventHubAsyncClient;
+        this.eventHubName = eventHubName;
+        this.consumerGroupName = consumerGroupName;
+        this.ownerId = ownerId;
+        this.inactiveTimeLimitInSeconds = inactiveTimeLimitInSeconds;
+        this.partitionPumpManager = partitionPumpManager;
+    }
+
+    /**
+     * This is the main method responsible for load balancing. This method is expected to be invoked by the {@link
+     * EventProcessor} periodically. Every call to this method will result in this {@link EventProcessor} owning <b>at
+     * most one</b> new partition.
+     * <p>
+     * The load is considered balanced when no active EventProcessor owns 2 partitions more than any other active
+     * EventProcessor.Given that each invocation to this method results in ownership claim of at most one partition,
+     * this algorithm converges gradually towards a steady state.
+     * </p>
+     * When a new partition is claimed, this method is also responsible for starting a partition pump that creates an
+     * {@link EventHubAsyncConsumer} for processing events from that partition.
+     */
+    public void loadBalance() {
+        /*
+         * Retrieve current partition ownership details from the datastore.
+         */
+        final Mono<Map<String, PartitionOwnership>> partitionOwnershipMono = partitionManager
+            .listOwnership(eventHubName, consumerGroupName)
+            .timeout(Duration.ofSeconds(1)) // TODO: configurable by the user
+            .collectMap(PartitionOwnership::partitionId, Function.identity());
+
+        /*
+         * Retrieve the list of partition ids from the Event Hub.
+         */
+        final Mono<List<String>> partitionsMono = eventHubAsyncClient
+            .getPartitionIds()
+            .timeout(Duration.ofSeconds(1)) // TODO: configurable
+            .collectList();
+
+        Mono.zip(partitionOwnershipMono, partitionsMono)
+            .map(this::loadBalance)
+            // if there was an error, log warning and TODO: call user provided error handler
+            .doOnError(ex -> logger.warning("Load balancing for event processor failed - {}", ex.getMessage()))
+            .subscribe();
+    }
+
+    /*
+     * This method works with the given partition ownership details and Event Hub partitions to evaluate whether the
+     * current Event Processor should take on the responsibility of processing more partitions.
+     */
+    private Mono<Void> loadBalance(final Tuple2<Map<String, PartitionOwnership>, List<String>> tuple) {
+
+        Map<String, PartitionOwnership> partitionOwnershipMap = tuple.getT1();
+        List<String> partitionIds = tuple.getT2();
+
+        if (ImplUtils.isNullOrEmpty(partitionIds)) {
+            // ideally, should never happen.
```

Comment:
"Ideally"  or "under normal circumstances"...  What I'm asking is whether this is something that would potentially occur and is expected or is an indicator that something is wrong.  The comment implies the former, but the implementation seems to imply the latter.

Reply:
Assuming this is reading the actual Event Hub metadata, this is impossible and would indicate a service failure of some sort.  Based on that, I think the comment is the artifact that is out of sync here.